spring-integration-file/src/main/java/org/springframework/integration/file/filters/AbstractMarkerFilePresentFileListFilter.java

```diff
@@ -0,0 +1,141 @@
+		this(filter, defaultFileNameFunction(suffix));
+	}
+
+	/**
+	 * Construct an instance with a single {@link FileNameFileListFilter} and a function
+	 * that will be applied to the name of a matched file when looking for the marker
+	 * file. The function returns the name of the marker file to match, or {@code null}
+	 * for never match. If a file {@code foo.txt} is matched by the filter and the
+	 * function returns "foo.txt.complete", this filter will only pass "foo.txt" if
+	 * "foo.txt.complete" is present.
+	 * @param filter the file name filter.
+	 * @param function the function to create the marker file name from the file name.
+	 */
+	public AbstractMarkerFilePresentFileListFilter(FileNameFileListFilter<F> filter,
+			Function<String, String> function) {
+		this(Collections.singletonMap(filter, function));
+	}
+
+	/**
+	 * Construct an instance with a map of {@link FileNameFileListFilter} and functions be
+	 * applied to the name of a matched file when looking for the marker file. i.e. if a
+	 * file {@code foo.txt} is matched by one of the filters and the corresponding
+	 * function returns "foo.txt.complete", this filter will only pass "foo.txt" if
+	 * "foo.txt.complete" is present. The function returns the name of the marker file to
+	 * match, or {@code null} for never match. Due to type erasure, we cannot provide a
+	 * constructor taking {@code Map<Filter, suffix}. For convenience, you can use
+	 * {@link #defaultFileNameFunction(String)} to use the default function used by the
+	 * {@link #AbstractMarkerFilePresentFileListFilter(FileNameFileListFilter, String)}
+	 * constructor.
+	 * @param filtersAndFunctions the filters and functions.
+	 */
+	public AbstractMarkerFilePresentFileListFilter(
+			Map<FileNameFileListFilter<F>, Function<String, String>> filtersAndFunctions) {
+		this.filtersAndFunctions.putAll(filtersAndFunctions);
+	}
+
+	/**
+	 * The default function used to create the file name for the corresponding marker file.
+	 * Appends a suffix to the file name.
+	 * @param suffix the suffix to append.
+	 * @return the function.
+	 */
+	public static Function<String, String> defaultFileNameFunction(final String suffix) {
+		return s -> s + suffix;
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public List<F> filterFiles(F[] files) {
+		if (files.length < 2) {
+			return Collections.emptyList();
+		}
+		final Set<String> candidates = Arrays.asList(files).stream()
+				.map(f -> getFilename(f))
+				.collect(Collectors.toSet());
+		List<F> results = new ArrayList<>();
+		for (F file : files) {
+			if (this.filtersAndFunctions.entrySet().stream().anyMatch(entry -> {
+				F[] fileToCheck = (F[]) Array.newInstance(file.getClass(), 1);
+				fileToCheck[0] = file;
+				if (entry.getKey().filterFiles(fileToCheck).size() > 0) {
+					String markerName = entry.getValue().apply(getFilename(file));
+					return markerName != null && candidates.contains(markerName);
+				}
+				return false;
+			})) {
```

Comment:
Just nit-picking from my head. Would look better with a local variable for the `anyMatch()` result.